src/extension/extensionServer.ts

```diff
@@ -0,0 +1,141 @@
+
+import * as em from "../common/extensionMessaging";
+import {FileSystem} from "../common/node/fileSystem";
+import {Packager} from "../common/packager";
+import {Log} from "../common/log";
+import * as Q from "q";
+import * as net from "net";
+import * as vscode from "vscode";
+
+
+export class ExtensionServer implements vscode.Disposable {
+    private outputChannel: any;
+    private serverInstance: net.Server = null;
+    private messageHandlerDictionary: { [id: number]: ((...argArray: any[]) => Q.Promise<any>) } = {};
+    private reactNativePackager: Packager;
+
+    public constructor(reactNativePackager: Packager) {
+        this.outputChannel = vscode.window.createOutputChannel("React-Native");
+        this.reactNativePackager = reactNativePackager;
+
+        /* register handlers for all messages */
+        this.messageHandlerDictionary[em.ExtensionMessage.START_PACKAGER] = this.startPackager;
+        this.messageHandlerDictionary[em.ExtensionMessage.STOP_PACKAGER] = this.stopPackager;
+        this.messageHandlerDictionary[em.ExtensionMessage.PREWARM_BUNDLE_CACHE] = this.prewarmBundleCache;
+    }
+
+    /**
+     * Starts the server.
+     */
+    public setup(): Q.Promise<void> {
+
+        let deferred = Q.defer<void>();
+
+        let launchCallback = (error: any) => {
+            Log.logMessage("Server started." + (error ? error : ""));
+            if (error) {
+                deferred.reject(error);
+            } else {
+                deferred.resolve(null);
+            }
+        };
+
+        this.serverInstance = net.createServer(this.handleSocket.bind(this));
+        this.serverInstance.on("error", this.recoverServer.bind(this));
+        this.serverInstance.listen(em.getPipePath(), launchCallback);
+
+        return deferred.promise;
+    }
+
+    /**
+     * Stops the server.
+     */
+    public dispose(): void {
+        if (this.serverInstance) {
+            this.serverInstance.close();
+            this.serverInstance = null;
+        }
+    }
+
+    /**
+     * Message handler for START_PACKAGER.
+     */
+    private startPackager(): Q.Promise<any> {
+        return this.reactNativePackager.start(this.outputChannel);
+    }
+
+    /**
+     * Message handler for STOP_PACKAGER.
+     */
+    private stopPackager(): Q.Promise<any> {
+        return this.reactNativePackager.stop(this.outputChannel);
+    }
+
+    /**
+     * Message handler for PREWARM_BUNDLE_CACHE.
+     */
+    private prewarmBundleCache(platform: string): Q.Promise<any> {
+        return this.reactNativePackager.prewarmBundleCache(platform);
+    }
+
+    /**
+     * Extension message handler.
+     */
+    private handleExtensionMessage(messageWithArgs: em.MessageWithArguments): Q.Promise<any> {
+        let handler = this.messageHandlerDictionary[messageWithArgs.message];
+        if (handler) {
+            Log.logMessage("Handling message: " + em.ExtensionMessage[messageWithArgs.message]);
```

Comment:
If this is intended to be an internal message, we should call logInternalMessage instead